pytorch_lightning/loops/optimizer/optimizer_loop.py

```diff
@@ -0,0 +1,376 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from copy import copy
+from functools import partial
+from typing import Any, Callable, Dict, List, Optional, Tuple
+
+import torch
+from torch import Tensor
+from torch.optim import Optimizer
+
+from pytorch_lightning.core.optimizer import LightningOptimizer
+from pytorch_lightning.loops import Loop
+from pytorch_lightning.loops.closure import Closure, ClosureResult
+from pytorch_lightning.loops.utilities import (
+    _block_parallel_sync_behavior,
+    _build_training_step_kwargs,
+    _check_training_step_output,
+    _process_training_step_output,
+    check_finite_loss,
+)
+from pytorch_lightning.trainer.connectors.logger_connector.result import ResultCollection
+from pytorch_lightning.trainer.progress import OptimizationProgress
+from pytorch_lightning.utilities import AMPType, AttributeDict, DeviceType, grad_norm
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.imports import _TPU_AVAILABLE
+
+_OUTPUTS_TYPE = List[List[Optional[ResultCollection]]]
+
+
+class OptimizerLoop(Loop):
+    """Runs over a sequence of optimizers. This loop implements what is known in Lightning as Automatic Optimization."""
+
+    def __init__(self):
+        super().__init__()
+        # TODO: use default dict here to simplify logic in loop
+        self.outputs: _OUTPUTS_TYPE = []
+        self.optim_progress: OptimizationProgress = OptimizationProgress()
+
+        self._skip_backward: bool = False
+        self._batch_idx: Optional[int] = None
+        self._optimizers: Optional[List[Optimizer]] = None
+        self._hiddens: Optional[Any] = None
+
+    @property
+    def done(self) -> bool:
+        """Returns ``True`` when the last optimizer in the sequence has run."""
+        return self.optim_progress.optimizer_idx >= len(self._optimizers)
+
+    def connect(self, **kwargs: "Loop") -> None:
+        raise NotImplementedError(f"{self.__class__.__name__} does not connect any child loops.")
+
+    def reset(self) -> None:
+        if not self.restarting:
+            self.optim_progress.optimizer_idx = 0
+        self.outputs = [[] for _ in range(len(self.trainer.optimizers))]
+
+    def on_run_start(self, batch: Any, hiddens: Any, optimizers: List[Optimizer], batch_idx: int) -> None:
+        self._batch_idx = batch_idx
+        self._optimizers = optimizers.copy()
```

Comment:
Why do we need to `copy`?

Reply:
we don't anymore :) removed.

the initial approach I had was to pop from the list every advance.